File: python/example10_3.py
# Example 10.3 Cochrane-Orcutt Iterative Procedure

import pandas as pd
import numpy as np
import statsmodels.api as sm

data = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/cjx.txt', sep='\s+', nrows=39)

X = np.log(data.X)
L = np.log(data.L1)
K = np.log(data.K1)
Z = pd.concat([L,K],axis=1)
Z = sm.add_constant(Z)
model_ols = sm.OLS(X,Z).fit()

# cochrane-orcutt / prais-winsten with given AR(1) rho, 
# derived from ols model, default to cochrane-orcutt 
def ols_ar1(model,rho,drop1=True):
    x = model.model.exog
    y = model.model.endog
    ystar = y[1:]-rho*y[:-1]
    xstar = x[1:,]-rho*x[:-1,]
    if drop1 == False:
        ystar = np.append(np.sqrt(1-rho**2)*y[0],ystar)
        xstar = np.append([np.sqrt(1-rho**2)*x[0,]],xstar,axis=0)
    model_ar1 = sm.OLS(ystar,xstar).fit()
    return(model_ar1)

# cochrane-orcutt / prais-winsten iterative procedure
# default to cochrane-orcutt (drop1=True)
def OLSAR1(model,drop1=True):
    x = model.model.exog
    y = model.model.endog
    e = y - (x @ model.params)
    e1 = e[:-1]; e0 = e[1:]
    rho0 = np.dot(e1,e[1:])/np.dot(e1,e1)
    rdiff = 1.0
    while(rdiff>1.0e-5):
        model1 = ols_ar1(model,rho0,drop1)
        e = y - (x @ model1.params)
        e1 = e[:-1]; e0 = e[1:]
        rho1 = np.dot(e1,e[1:])/np.dot(e1,e1)
        rdiff = np.sqrt((rho1-rho0)**2)
        rho0 = rho1
        print('Rho = ', rho0)
    # pint final iteration
    # print(sm.OLS(e0,e1).fit().summary())
    model1 = ols_ar1(model,rho0,drop1)
    return(model1)

# AR(1) based on cochrane-orcutt iterative procedure   
ar1_co = OLSAR1(model_ols)
# ar1_co = OLSAR1(model_ols,drop1=True)
print(ar1_co.summary())

# AR(1) based on prais-winsten iterative procedure
ar1_pw = OLSAR1(model_ols,drop1=False)
print(ar1_pw.summary())
# the results are based on transformed model

# alternatively, using statsmodels' GLSAR to estimate AR(!)
# results may not be the same as OLSAR, need to check
ar1_gls = sm.GLSAR(X,Z,1)
results = ar1_gls.iterative_fit(maxiter=50)
print ('Iterations used = %d Converged %s' % (results.iter, results.converged) )
print ('Rho =  ', ar1_gls.rho)
print(results.summary())

# alternatively, ARIMA model may be used
# results may not be the same as OLSAR and GLSAR, need to check
from statsmodels.tsa.arima_model import ARMA
ar1_ml = ARMA(X,exog=Z,order=(1,0)).fit(trend='nc',method='mle')
print(ar1_ml.summary())

File: R/example10_3.R
# Example 10.3 Cochrane-Orcutt Iterative Procedure

setwd("C:/Course17/ceR/R")
data<-read.table("http://web.pdx.edu/~crkl/ceR/data/cjx.txt",header=T,nrows=39)

year <- data$YEAR
X <- log(data$X)
L <- log(data$L1)
K <- log(data$K1)
results.ols<-lm(X~L+K)
summary(results.ols)

# cochrane-orcutt lm, given AR(1) rho, derived from ols model
lm.ar1.co <- function(model,rho) {
  x <- model.matrix(model)
  y <- model.response(model.frame(model))
  n <- length(y)
  t <- 2:n
  ystar <- y[t] - rho * y[t-1]
  xstar <- x[t,] - rho * x[t-1,]
  return(lm(ystar ~ xstar - 1))
}

# prais-winsten lm given AR(1) rho, derived from ols model
lm.ar1.pw <- function(model,rho) {
  x <- model.matrix(model)
  y <- model.response(model.frame(model))
  n <- length(y)
  t <- 2:n
  ystar <- y[t] - rho * y[t-1]
  xstar <- x[t,] - rho * x[t-1,]
  ystar <- c(sqrt(1-rho^2)*y[1],ystar)
  xstar <- rbind(sqrt(1-rho^2)*x[1,],xstar)
  return(lm(ystar ~ xstar - 1))
}

# cochrane-orcutt procedure
lm.co<-function(model) {
  x<-model.matrix(model)
  y<-model.response(model.frame(model))
  e<-y-x%*%model$coefficients
  e1<-c(NA,e[1:length(e)-1])
  r0<-lm(e~e1-1)$coefficients
  rdiff<-1
  while(rdiff>1.0e-5) {
    model1<-lm.ar1.co(model,r0)
    e<-y-x%*%model1$coefficients
    e1<-c(NA,e[1:length(e)-1])
    rho<-lm(e~e1-1)$coefficients
    rdiff<-sum((rho-r0)^2)
    r0<-rho
  }
  list(lm.co=summary(lm.ar1.co(model,r0)),rho=r0)
}
results.co<-lm.co(results.ols) # the results are based on transformed model
results.co

# prais-winsten procedure
lm.pw<-function(model) {
  x<-model.matrix(model)
  y<-model.response(model.frame(model))
  e<-y-x%*%model$coefficients
  e1<-c(NA,e[1:length(e)-1])
  r0<-lm(e~e1-1)$coefficients
  rdiff<-1
  while(rdiff>1.0e-5) {
    model1<-lm.ar1.pw(model,r0)
    e<-y-x%*%model1$coefficients
    e1<-c(NA,e[1:length(e)-1])
    rho<-lm(e~e1-1)$coefficients
    rdiff<-sum((rho-r0)^2)
    r0<-rho
  }
  list(lm.pw=summary(lm.ar1.pw(model,r0)),rho=r0)
}
results.pw<-lm.pw(results.ols) # the results are based on transformed model
results.pw

# alternatively, using package orcutt, prais
library(orcutt)
results1 <- cochrane.orcutt(results.ols)
results1

library(prais)
results2<-prais.winsten(results.ols,data=NULL)
results2

library(nlme)
results3<-gls(X~L+K,corr=corAR1(),method="ML")
results3

rm(list=ls())
